### ISSUE TITLE
Add support for reading extended DLL characteristics to PEReader

### ISSUE THREAD

**Issue by gsuberland** [dotnet/runtime#51839]
## Background and Motivation

The [PE file format](https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#debug-type) may include a debug directory entry of the type `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS`, which contains characteristic flags in addition to those included in the `DllCharacteristics` field of the optional header. This extension allows for more characteristic bits to be set than can be reasonably stored in the original field.

At current the only defined bit in there is `IMAGE_DLLCHARACTERISTICS_EX_CET_COMPAT` (0x0001), which specifies that the binary has been compiled with support for Intel Control-flow Enforcement Technology (CET), also known as Shadow Stacks, which is used in the [Hardware-enforced Stack Protection (HSP)](https://techcommunity.microsoft.com/t5/windows-kernel-internals/understanding-hardware-enforced-stack-protection/ba-p/1247815) security feature.

While the documentation of this particular debug structure is somewhat lacking, with no information about the exact structure of this entry beyond the supported flags, it appears to begin with a 16-bit extended flags field, with contents as documented in the PE file format document linked above. This tallies with my practical investigation of executables built with this particular feature. The MSVC compiler uses the [`/CETCOMPAT`](https://docs.microsoft.com/en-us/cpp/build/reference/cetcompat?view=msvc-160) flag to turn support on, which in turn sets this flag. I believe there may be more structures added to the binary when this feature is turned on, to define exception handler continuation targets, CET enforced ranges, and dynamic ranges, but I have not studied the feature in sufficient detail to fully understand the details. Ideally the documentation would be expanded to include this. Anyway, this is rather orthogonal to this issue.

The `PEReader` class currently has support for parsing certain types of debug records, but not this one. As the extended characteristics field is likely to see more use in the future, particularly in relation to security features, it would be helpful to add support for reading it.

I've opened this issue as it relates to some ongoing research I'm doing into the interoperability and compatibility of Windows exploit mitigation features and .NET Core / .NET 5 executables. I believe it may make sense to take greater advantage over these features when executables are being run on a supported platform, particularly via opt-in flags in app configs / manifests. @blowdart asked that I tag him in these issues - hi, Barry! I believe that having greater visibility of these features' associated PE structures is a good place to lay the groundwork.

## Proposed API

The API would mirror that of the CodeView debug data API, for example:

```csharp
public ExtendedDllCharacteristicsDebugDirectoryData ReadExtendedDllCharacteristics(DebugDirectoryEntry entry);
```

The `ExtendedDllCharacteristicsDebugDirectoryData` class would be implemented in much the same was as `CodeViewDebugDirectoryData`. The exact contents of the class would be dependent on any internal Microsoft documentation regarding the complete structure of this entry, but it would at minimum include a flags enum for the defined extended characteristics - probably just CET Compatible for now.

A new enum value would be added to `DebugDirectoryEntryType`, matching `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS`.

## Usage Examples

Usage should be fairly self-evident, but here is an example for completeness:

```csharp
bool IsImageCetCompatible(PeReader peReader)
{
  var extendedCharacteristicsDirectory = 
    peReader.ReadDebugDirectory()
      .FirstOrDefault(dd => dd.Type == DebugDirectoryEntryType.ExtendedDllCharacteristics);
  if (extendedCharacteristicsDirectory != null)
  {
    var extendedCharacteristics = peReader.ReadExtendedDllCharacteristics(extendedCharacteristicsDirectory);
    return extendedCharacteristics.CetCompatible;
  }
  return false;
}
```

## Risks

No risks that I can think of - this does not make any modifications that would affect code already in production.

**Comment by blowdart** [comment]
@jaredpar @GrabYourPitchforks for thoughts on this defence in depth in a future version.

**Comment by jkotas** [comment]
We have work in progress on making CoreCLR compatible with CET. This work is tracked by https://github.com/dotnet/runtime/issues/47309 . cc @kouvel 

Adding convenience System.Reflection.Metadata APIs for `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` should not be needed to make CoreCLR work well with CET.

**Comment by gsuberland** [comment]
> Adding convenience System.Reflection.Metadata APIs for `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` should not be needed to make CoreCLR work well with CET.

My thought with this specific issue wasn't as much about getting CoreCLR to work with CET, although it's excellent to hear that this is in the pipeline, but more so that .NET applications that utilise the PE metadata API could conveniently and reliably access the extended characteristics flags in order to identify CET support (and other feature flags in future). This is applicable in tooling such as [binskim](https://github.com/microsoft/binskim), which I plan to open a PR for (to add CET coverage) if this API enhancement is accepted.

**Comment by jkotas** [comment]
The PE reader in System.Reflection.Metadata has the convenience helper methods tp decode the small subset of PE file structuctures that are relevant for managed binaries.

binskim has a decoder for whatever additional it needs. For example, it has a wrapper for `IMAGE_LOAD_CONFIG_DIRECTORY32` in https://github.com/microsoft/binskim/blob/main/src/BinaryParsers/PEBinary/PortableExecutable/ImageLoadConfigDirectory32.cs . It can have wrapped just like this one for `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` flags. We can still add this API if we decided that it belongs to System.Reflection.Metadata. I am just saying that this API should not be required for binskim to make progress.

I believe @eddynaka is working on adding support for CET right now. @eddynaka do you believe that you can fix https://github.com/microsoft/binskim/issues/277 without any new APIs in System.Reflection.Metadata?

**Comment by eddynaka** [comment]
@jkotas , yes. I'm working on that. Now, I'm not sure how similar the `IMAGE_LOAD_CONFIG_DIRECTORY32` is to `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS`.

Do you know how that works?

**Comment by jkotas** [comment]
`IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` is much simpler. It is just a single Int16.

**Comment by gsuberland** [comment]
From what I've read, the two structures have separate functions in terms of indicating CET support.

The lowest bit of the `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` Int16 specifies whether CET is supported by the binary at all. This is set when `/CETCOMPAT` is specified as an MSVC linker flag.

The image load config directory contains, among other things, a lot of RVAs and counts for tables relating to CFG. The `GuardFlags` field has a number of flags to indicate support for various CFG features, including `IMAGE_GUARD_EH_CONTINUATION_TABLE_PRESENT` (0x200000) which specifies that CET EHCont was enabled via the `/guard:ehcont` flag. The config directory also contains the RVA and count fields that describe the exception handling continuation metadata table - `GuardEHContinuationTable` and `GuardEHContinuationCount`.

As I understand it, an image may have the `IMAGE_DLLCHARACTERISTICS_EX_CET_COMPAT` flag set without EHCont being enabled, if only `/CETCOMPAT` is set but `/guard:ehcont` is not (in MSVC, at least), so just checking for EH continuation table presence is not a complete check for an image's CET support.

**Comment by steveharter** [comment]
This doesn't seem necessary -- above [binskim](https://github.com/microsoft/binskim) is recommended.

**Comment by gsuberland** [comment]
@steveharter The point of me mentioning binskim was that this feature would be useful _in_ binskim, not that binskim is a useful replacement for this feature.

**Comment by jkotas** [comment]
> The point of me mentioning binskim was that this feature would be useful in binskim

I agree with that, but that it is not the right justification for adding this decoder to System.Reflection.Metadata.

Many other datastructures from PE file format spec are not in System.Reflection.Metadata either. System.Reflection.Metadata contains only the small subset of PE file format spec that is directly relevant to managed code. Extended DLL characteristics are not directly relevant to managed code.

**Comment by gsuberland** [comment]
I can see your point, but the other side of the coin is that Microsoft sets the PE standard, so who better than to provide the appropriate parsing functionality within .NET? One could also make the point that C++/IL binaries and related compiler & test code would benefit from being able to parse the native structures.

**Comment by jkotas** [comment]
> Microsoft sets the PE standard, so who better than to provide the appropriate parsing functionality within .NET?

Nothing says that there has to be fully featured PE file parser included in the core .NET libraries. There is not a standard PE file parser library for C++ either. Parsing of native unmanaged binary formats is a very niche scenario. It does not make sense to include in the core .NET libraries.

> Microsoft's security team are actively producing tooling for scanning binaries

This tooling has been adding their own local parser for additional PE file structures as needed without problems: https://github.com/microsoft/binskim/tree/main/src/BinaryParsers/PEBinary

**Comment by teo-tsirpanis** [comment]
Here's a concrete proposal, if we want to move this forward. I'm not asking for this API myself.

```csharp
namespace System.Reflection.PortableExecutable;

public class DebugDirectoryBuilder
{
    public void AddExtendedDllCharacteristicsEntry(ExtendedDllCharacteristics characteristics);
}

public enum DebugDirectoryEntryType
{
    ExtendedDllCharacteristics = 20,
}

[Flags]
public enum ExtendedDllCharacteristics : ushort
{
    None = 0,
    CetCompatible = 0x0001,
    ForwardCfiCompatible = 0x0040,
}

public readonly struct ExtendedDllCharacteristicsDebugDirectoryData
{
    public ExtendedDllCharacteristics ExtendedDllCharacteristics { get; }
}

public class PEReader
{
    public ExtendedDllCharacteristicsDebugDirectoryData ReadExtendedDllCharacteristicsDebugDirectoryData(DebugDirectoryEntry entry);
}
```

**Comment by jkotas** [comment]
My suggestion is to close this issue. I do not think that this surface belongs to System.Reflection.Metadata.
